pkg/util/tls/tls.go

```diff
@@ -13,42 +13,57 @@ import (
 
 // ClientConfig is the config for client TLS.
 type ClientConfig struct {
-	CertPath string `yaml:"tls_cert_path"`
-	KeyPath  string `yaml:"tls_key_path"`
-	CAPath   string `yaml:"tls_ca_path"`
+	CertPath           string `yaml:"tls_cert_path"`
+	KeyPath            string `yaml:"tls_key_path"`
+	CAPath             string `yaml:"tls_ca_path"`
+	ServerName         string `yaml:"tls_server_name"`
+	InsecureSkipVerify bool   `yaml:"tls_insecure_skip_verify"`
 }
 
 // RegisterFlagsWithPrefix registers flags with prefix.
 func (cfg *ClientConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.StringVar(&cfg.CertPath, prefix+".tls-cert-path", "", "TLS cert path for the client")
 	f.StringVar(&cfg.KeyPath, prefix+".tls-key-path", "", "TLS key path for the client")
 	f.StringVar(&cfg.CAPath, prefix+".tls-ca-path", "", "TLS CA path for the client")
+	f.StringVar(&cfg.ServerName, prefix+".tls-server-name", "", "Override the expected name on the server certificate")
+	f.BoolVar(&cfg.InsecureSkipVerify, prefix+".tls-insecure-skip-verify", false, "TLS skip certificate validation")
 }
 
 // GetTLSConfig initialises tls.Config from config options
 func (cfg *ClientConfig) GetTLSConfig() (*tls.Config, error) {
-	if cfg.CertPath != "" && cfg.KeyPath != "" && cfg.CAPath != "" {
-		clientCert, err := tls.LoadX509KeyPair(cfg.CertPath, cfg.KeyPath)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to load TLS certificate %s,%s", cfg.CertPath, cfg.KeyPath)
-		}
+	// no tls config given at all
+	if cfg.CertPath == "" && cfg.KeyPath == "" && cfg.CAPath == "" && !cfg.InsecureSkipVerify && cfg.ServerName == "" {
```

Comment:
Do you think it's worth to also check for `&& !cfg.InsecureSkipVerify && cfg.ServerName == ""` or we could just assume that if no cert file has been configured then TLS is disabled?

Reply:
I think that check is there for the case when someone what's to connect via TLS, but is not using client certificate auth.

I think it is still valid to skipVerify all together or make use of the ServerName flag to require a certain hostname on the server ceritficate.

Reply:
> I think that check is there for the case when someone what's to connect via TLS, but is not using client certificate auth.

Is this case supported? Have you ever tested it? Also, following the case you describe, looks there's no way to use it unless disabling skip verify or setting the server name.

Reply:
It would be clearer to me to have special flag to enable TLS. I'm not sure if just setting `InsecureSkipVerify` should enable TLS. WDYT?

For example, it should be enough to just *enable* TLS via this flag, without setting anything else: in that case default set of CA certificates is used, with no client authentication. I think that should be a valid option, but right now that's not possible.

Reply:
I'm have a private conversation with Christian about this. I agree with you Peter, but I believe it's out of the scope of this PR honestly. This PR was supposed a bug fix and it's turning out to be a feature / change as well.

Reply:
My take is:

This is our code:
```
func (cfg *ClientConfig) GetGRPCDialOptions() ([]grpc.DialOption, error) {
	if tlsConfig, err := cfg.GetTLSConfig(); err != nil {
		return nil, errors.Wrap(err, "error creating grpc dial options")
	} else if tlsConfig != nil {
		return []grpc.DialOption{grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))}, nil
	}
	return []grpc.DialOption{grpc.WithInsecure()}, nil
}
```

If `cfg.GetTLSConfig()` returns `nil`, we do intentionally disable it `grpc.WithInsecure()`. My whole point is:
- You can't enable TLS (with no certs) without specifying either `InsecureSkipVerify` or `ServerName`, which looks weird to me

This looks a bug to me, but looks a bug introduced by this PR because before TLS was only meant to cover authentication (and so necessarily required certificates). In particular:
- Before this PR: TLS was meant for authentication, so requiring to always provide certs
- In this PR: TLS can be used for encryption as well, but you can't enable it without specifying either `InsecureSkipVerify` or `ServerName`

Reply:
> Before this PR: TLS was meant for authentication, so requiring to always provide certs

1) Client verifies server, and 2) server optionally verifies client. My point is that for 1) you should not need to provide anything, and use root certificates provided by the host (Go runtime already tries hard to find them). For 2), you obviously need to provide client certificate.

When I look at changelog, we haven't announced this feature as client-only authentication, but as "TLS support". 

> In this PR: TLS can be used for encryption as well, but you can't enable it without specifying either InsecureSkipVerify or ServerName

This is obviously wrong. But always requiring CA certs OR client certs is wrong too.

Reply:
I think we're all on the same page, but with different takes on what to do. Let's try to come up with action items.

Some facts:
1. The version in `master` requires certs to enable TLS
2. This PR was initially meant to be a bugfix
3. This PR is moving forward improving TLS, allowing to use TLS for "1) Client verifies server" only use case, but it's a bit buggy because you can't enable use case (1) without either disabling verification or setting server name

The question we have to answer is:
- Do we want this PR to be a bugfix or a mix of bugfix + additional feature?

My general suggestion is to keep separation of concerns so, if it's to me, I would:
- Keep this PR as bug fix (no logic change except for the skip verification)
- Open a new PR with the TLS improvements

If you prefer to do both in this PR I'm OK as well, but I believe we should fix the config issue. I guess we should add an explicit option to enable TLS.

Reply:
The main reason this became an "enhancement" as well, was that the integration tests only work if you either `skip-verify` or provide a `server-name`. Otherwise there will be a validation error, as we are connecting to unpredictable hosts IP addresses, which we get from consul (I assume).

If we want to keep it as minimal as possible we probably should set skip-verify to true in the integration tests and then iterate on the whole TLS implementation. 

My feeling is we need to treat GRPC and HTTP client configuration slightly different:

If you enable TLS with HTTP you are required to use https:// in the URL explicitly. For GRPC we default to tls-insecure-skip-verify being true, through this https://github.com/cortexproject/cortex/blob/a4aad5da5e4ea05110d7ff389e736b62ecbe6f34/pkg/util/tls/tls.go#L61. If we remove that check we would require valid certificates for everyone.

Reply:
> If we want to keep it as minimal as possible we probably should set skip-verify to true in the integration tests and then iterate on the whole TLS implementation.

Setting `skip-verify` in integration tests sounds reasonable to me.

> If you enable TLS with HTTP you are required to use https:// in the URL explicitly.

Right. It's not so much about enabling rather than configuring TLS in this case.

> For GRPC we default to tls-insecure-skip-verify being true, through this

Whether TLS is used for gRPC or not should be specified for each gRPC config, and default to false (and when using `grpc.WithInsecure()`). It looks like it should not even be in `tls.ClientConfig` config then. 🤔

Reply:
To move further with this PR, I believe Marco's suggestion is to do

```suggestion
	if cfg.CertPath == "" && cfg.KeyPath == "" && cfg.CAPath == "" {
```

i.e. to enable TLS only if certificates are configured. That is compatible with behavior on master, although not as clear as explicit flag, but we can do the flag in other PR.

Reply:
I will split out the bugfix from the enhancement of allowing a ServerName specified. 

Without having the ability to configure a expected ServerName, we will not have a feasible way of running running cortex without skipping verify (as explained in my comment before), so it should remain a priority sorting that out. We will need to think a bit more how to expose that flag for GRPC clients. Another big focus needs to be testing TLS. The current coverage is quite minimal and will even be more limited by skipping verify.